### PR TITLE
Improve the DMA sniffer's API (#887)

### LINKED ISSUE #887: Add API access to DMA SNIFF_CTRL bits OUT_INV and OUT_REV
This came up in the discussion of raspberrypi/pico-feedback#247. AFAICT there is no API in the SDK for accessing these two bits in DMA's `SNIFF_CTRL` register, so they have to be set by hand in `dma_hw`.

It turns out that by using them, it is possible for the sniffer to generate a CRC32 value that matches the value returned by zlib's crc32() function, making it possible to automate data integrity checks after transferring data from a Pico to a host. So a convenience function in the SDK would be valuable.

How that was done is spelled out in the pico-feedback issue, and in a thread on the Pico/General forum:

https://forums.raspberrypi.com/viewtopic.php?t=336582

### PR DESCRIPTION
**Pull request by andygpz11**
The DMA sniffer's API did not include all the functionality required to "sniff" DMA transfers.  

Add 2 new functions and improve the existing sniffer enable function so that it does not clear the bits it doesn't touch.  Fixes #887

### CODE REVIEW

File: src/rp2_common/hardware_dma/include/hardware/dma.h [hunk 1]
```diff
@@ -761,6 +765,34 @@ inline static void dma_sniffer_set_byte_swap_enabled(bool swap) {
         hw_clear_bits(&dma_hw->sniff_ctrl, DMA_SNIFF_CTRL_BSWAP_BITS);
 }
 
+/*! \brief Enable the Sniffer output invert function
+ *  \ingroup hardware_dma
+ *
+ * Perform a bit inversion on the sniffed data output
```
Comment (kilograham) [inline src/rp2_common/hardware_dma/include/hardware/dma.h:771]: tense is a bit weird here, I would go with "If enabled, the sniff data result appears bit-reversed when read. This does not affect the way the checksum is calculated" which is roughly what the registers say

File: src/rp2_common/hardware_dma/include/hardware/dma.h [hunk 2]
```diff
@@ -761,6 +765,34 @@ inline static void dma_sniffer_set_byte_swap_enabled(bool swap) {
...
+    if (invert)
+        hw_set_bits(&dma_hw->sniff_ctrl, DMA_SNIFF_CTRL_OUT_INV_BITS);
+    else
+        hw_clear_bits(&dma_hw->sniff_ctrl, DMA_SNIFF_CTRL_OUT_INV_BITS);
+}
+
+/*! \brief Enable the Sniffer output bit reversal function
+ *  \ingroup hardware_dma
+ *
+ * Perform a bit inversion on the sniffed data output
```
Comment (kilograham) [inline src/rp2_common/hardware_dma/include/hardware/dma.h:785]: same comment as above, and also this is bit reverse not bit inversion

**Review by kilograham — CHANGES_REQUESTED** [review]

File: src/rp2_common/hardware_dma/include/hardware/dma.h [hunk 3]
```diff
@@ -769,6 +801,27 @@ inline static void dma_sniffer_disable(void) {
     dma_hw->sniff_ctrl = 0;
 }
 
+/*! \brief Set the sniffer's data accumulator with initial value
+ *  \ingroup hardware_dma
+ *
+ * Generally, CRC algorithms are used with the data accumulator initially
+ * seeded with 0xFFFF or 0xFFFFFFFF (for ccrc16 and crc32 algorithms)
```
Comment (lurch) [inline src/rp2_common/hardware_dma/include/hardware/dma.h:808]: ccrc16 -> crc16 ?

File: src/rp2_common/hardware_dma/include/hardware/dma.h (hunk 1 above)
Comment (andygpz11) [reply src/rp2_common/hardware_dma/include/hardware/dma.h]: Thanks, agreed, resolved (I hope).

File: src/rp2_common/hardware_dma/include/hardware/dma.h (hunk 2 above)
Comment (andygpz11) [reply src/rp2_common/hardware_dma/include/hardware/dma.h]: Thanks, agreed, resolved (I hope).

File: src/rp2_common/hardware_dma/include/hardware/dma.h (hunk 3 above)
Comment (andygpz11) [reply src/rp2_common/hardware_dma/include/hardware/dma.h]: Thanks, agreed, resolved (I hope).

**Review by kilograham — APPROVED** [review]

**Pull request merged**